# 00 - pynhm Python Virtual Environments

*James McCreight, September 2022*

---

This is the first notebook to get started. We'll clone the git repository and setup the python virtual environment we need. __This notebook it is to be run with a bash kernel__ in jupyterlab (or jupyter notebook).

## Requirements: Git and Anaconda (or pip)
To follow this notebook, you'll need git and conda (Anaconda or Miniconda). You may elect to use pip instead of conda for installing dependencies, and there is a note below on this. However [installing Anaconda or Miniconda](https://docs.anaconda.com/anaconda/install/) is recommended. 

In fact, if you have installed conda, you bootstrap everything else you need to run this notebook (including git and jupyter) with the following code block:

```bash
env_name=git_jupyter
env_file=${env_name}.yml

echo "name: git_jupyter
channels:
  - conda-forge
dependencies:
  - git
  - bash_kernel
  - ipython
  - jupyter
  - jupyterlab  
" > $env_file

conda remove -y --name $env_name --all || exit 2
conda create -y --name $env_name || exit 3
conda env update --name $env_name --file $env_file --prune  || exit 4
rm $env_file || exit 5
```

In general, conda installs can take some time (which is the downpayment on headaches saved, i guess), so please be prepared to wait when running this and subsequent conda commands. 

In creating this notebook, the above code was run and then followed by

```bash
conda activate git_jupyter
jupyter-lab
```

From there, this notebook was opened and the rest of the code run interactively. You should be able to replicate this.

Once you have git, conda, pip, and Python, installed versions of this software should be "reasonably" close to what is printed below. 

In [1]:
git --version
conda --version
pip --version
python --version

git version 2.37.3
conda 4.14.0
pip 22.2.2 from /Users/jamesmcc/opt/anaconda3/envs/git_jupyter/lib/python3.10/site-packages/pip (python 3.10)
Python 3.10.6


The Python version installed by conda should be in the range 3.8 - 3.10. 

## Set pynhm_repo_root
Specify the root of the pynhm repository (still yet to be cloned below) here.

In [2]:
# pynhm_repo_root=/your/path/to/pynhm
pynhm_repo_root=/Users/jamesmcc/usgs/pynhm

## Clone the pynhm repository

In [3]:
cd $(dirname $pynhm_repo_root)
if [ ! -e pynhm ]; then
    git clone https://github.com/EC-USGS/pynhm
fi

If you'd like to have your own fork, please fork the repository on github and then clone your fork instead of the upstream (as above). 

## Download the GIS files
You can choose to put these elsewhere, but the location used here will be assumed in the subsequent notebooks.

## Setup the virtual environment

What are virtual environments? Virtual environments are like separate python toolboxes for different specific projects or tasks. When you need to work on a specific task, you load the associated virtual environment and use the specific tools (packages and versions) contained therein. You load and unload the virtual environments (from the command line or in jupyter) before or when starting Python. 

There are three virtual environments listed in the pynhm repository.

In [4]:
echo "pynhm_repo_root = $pynhm_repo_root"
ls ${pynhm_repo_root}/ci/requirements/environment.yml || echo "environment.yml not found, please check pynhm_repo_root"
ls ${pynhm_repo_root}/ci/requirements/doc.yml || echo "docs.yml not found, please check pynhm_repo_root"
ls ${pynhm_repo_root}/examples/examples_env.yml || echo "examples_env.yml not found, please check pynhm_repo_root"

pynhm_root = /Users/jamesmcc/usgs/pynhm
/Users/jamesmcc/usgs/pynhm/ci/requirements/environment.yml
/Users/jamesmcc/usgs/pynhm/ci/requirements/doc.yml


These YAML files are read by Anaconda to setup an environment with a requested name (as we will do below). Note that if you wanted to use pip instead of conda, you can use the `.txt` equivalents of these files. 

We install the `examples_env.yaml` below. The `environment.yml` environment is the minimal set of dependences for CI. If you are interested in building the documentation for the repository, you would want to install and use the `doc.yml` to do so. 

We take a look at `envionment.yml`

In [5]:
cat ${pynhm_repo_root}/examples/examples_env.yml

name: pynhm-tests
channels:
  - conda-forge
  - nodefaults
dependencies:
  - netCDF4
  - networkx
  - numpy
  - numba
  - pandas
  - pip
  - pytest
  - pytest-cov
  - pytest-env
  - pytest-xdist
  - pyyaml
  - xarray
  - flopy
  - pip:
      - click != 8.1.0
      - black
      - isort
      - flake8
      - pylint


We see that Anaconda will create an environment named "pynhm-tests", what conda channels it will search for dependencies, the dependencies themselves, and what dependencies are only available from pip. 

Conda allows us to install these into an environment of a different name, and we'll create the environment "pynhm_nb". We name it this because we'll also want to install IPython and jupyter, so we can run notebooks and pynhm. The `environment.yml` file tries to keep a minimal listing of pynhm dependencies. We may consider creating a `notebooks.yml` file which would contain the additional dependencies to run our notebooks in the future. 

When installing environments from conda, I prefer to delete and recreate from scratch. So the following code will completely delete the named environment ("pynhm_nb"). (Note the following will not succeed if the named environment is active). 

In [6]:
env_name=pynhm_nb

In [7]:
env_file=${pynhm_repo_root}/examples/examples.yml

echo "Removing $env_name..."
conda remove -y --name $env_name --all || exit 2

echo
echo "Creating $env_name..."
conda create -y --name $env_name || exit 3

echo
echo "Installing $env_name from $env_file"
time conda env update --name $env_name --file $env_file --prune  || exit 4

Removing pynhm_plus...

Remove all packages in environment /Users/jamesmcc/opt/anaconda3/envs/pynhm_plus:

No packages found in /Users/jamesmcc/opt/anaconda3/envs/pynhm_plus. Continuing environment removal

Creating pynhm_plus...
Solving environment: done

## Package Plan ##

  environment location: /Users/jamesmcc/opt/anaconda3/envs/pynhm_plus



Preparing transaction: done
Verifying transaction: done
Executing transaction: done
#
# To activate this environment, use
#
#     $ conda activate pynhm_plus
#
# To deactivate an active environment, use
#
#     $ conda deactivate

Retrieving notices: ...working... done

Installing pynhm_plus from /Users/jamesmcc/usgs/pynhm/examples/examples.yml
Solving environment: done

hvplot-0.8.1         | 3.0 MB    | ##################################### | 100% 
pyviz_comms-2.2.1    | 31 KB     | ##################################### | 100% 
datashader-0.14.2    | 16.6 MB   | ##################################### | 100% 
Preparing transaction: done
Verif

Now we need to install pynhm. 

In [8]:
# You cant activate conda in this bash shell, so use the CONDA_PATH to get around installing
conda_env_path=$(dirname $CONDA_PREFIX)
cd $pynhm_repo_root
${conda_env_path}/${env_name}/bin/pip install -e .

Obtaining file:///Users/jamesmcc/usgs/pynhm
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... done
  Preparing editable metadata (pyproject.toml) ... done
  Building editable for pynhm (pyproject.toml) ... done
  Created wheel for pynhm: filename=pynhm-0.1.dev479+g4a1ecbe.d20220902-0.editable-py3-none-any.whl size=3175 sha256=0bd0107830fe06d7499ea05bd7f8df5eda44244102b327c60d341fb659a3028d
  Stored in directory: /private/var/folders/rf/pj_9dt9x55b9gfnv34f3xktc0000gn/T/pip-ephem-wheel-cache-s15o8aa_/wheels/9a/f7/d8/7f6c6f238417a40f9afa97b355c56d869742bc25a749b3e278
Successfully built pynhm


We are now ready to use pynhm. If you're in a jupyter notebook in the git_jupyter kernel, shut down that notebook and that jupyter instance. In your shell, activate the new "pynhm_nb" environment and start jupyter-lab:

```bash
conda activate pynhm_nb
jupyter-lab
```

Once in jupyter-lab, open notebook 01 with the bash kernel "conda env: pynhm_nb".